_source/_posts/2022-04-05-state-of-ffi-java.md

```diff
@@ -0,0 +1,280 @@
+  - Initial support for C interop
+  - Call native code in a `.dll`, `.so`, or `.dylib`
+  - Create a native function pointer to a Java method that can be passed to code in a native library
+- JEP-389 - First incubator in JDK 16
+
+### Vector API
+
+Next is the vector API, which is crucial for FFI, especially in machine learning and advanced computations.
+
+- API for reliable and performant vector computations
+  - Platform agnostic
+  - Clear and concise API
+  - Reliable runtime compilation and performance
+  - Graceful degradations
+- JEP-338 - First incubator in JDK 16
+- JEP-414 - Second incubator in JDK 17
+- JEP-417 - Third incubator in JDK 18
+
+### Foreign Function & Memory API
+
+Finally, the Foreign Linker API & Foreign-Memory Access API has evolved together to become Foreign Function & Memory API. It was first incubated in JDK 17.
+
+- Evolution of the Foreign-Memory Access API and the Foreign Linker API
+  - Same goals and features as the original two (Ease of use, safety, performance, generality)
+- JEP-412 - First incubator in JDK 17
+- JEP-419 - Second incubator in JDK 18
+
+### Jextract
+
+And finally, there is the fantastic jextract tool. While it's not an API or part of the JDK itself, it is an essential tool for project Panama.
+
+- A simple command-line tool
+- Generates a Java API from one or more native C headers
+- Shipped with OpenJDK Panama builds at the moment and will be part of the JDK in the future
+- Makes working with large C headers a cakewalk
+
+For example, to generate a Java API for OpenGL with Jextract, you could simply run the following on a Unix like OS:
+
+```bash
+jextract --source -t org.opengl -I /usr/include /usr/include/GL/glut.h
+```
+
+## JNI vs. Panama
+
+Since JNI is the current standard and Panama aims to replace that, it makes sense to compare the two. Let's take a simple example of calling the `getpid` function from the standard C `unistd` header.
+
+### JNI
+
+{% img blog/state-of-ffi-java/JNI-getpid.png alt:"getpid with JNI" width:"800" %}{: .center-image }
+
+As you can see here, there are precisely six steps to make this simple native call using JNI. You start by writing a Java class that declares the native method. Then you use `javac` to generate a header file and a C class for this. These are the native bindings. Next, you will implement the C class. remember, these are Java developers writing C code. This means you must write memory-safe C code, which has access to the entire JVM via the `JNIEnv` variable passed to the C class. The developer probably does not have much experience in C in many scenarios. So that will be fun or more like a security nightmare. Next, you will compile the C code into a platform-specific dynamic library and determine where to place that binary. Then probably also pray that all this works without exposing the app to a security vulnerability. Next, you will load this into the Java class and compile and run the class, and hopefully, it works.
```

Comment:
```suggestion
As you can see here, there are precisely six steps to make this simple native call using JNI. You start by writing a Java class that declares the native method. Then you use `javac` to generate a header file and a C class for this. These are the native bindings. Next, you will implement the C class. Remember, these are Java developers writing C code. This means you must write memory-safe C code, which has access to the entire JVM via the `JNIEnv` variable passed to the C class.  In many scenarios, the developer may not have much experience in C. So that will be fun... or more like a security nightmare. Next, you will compile the C code into a platform-specific dynamic library and determine where to place that binary. Pray that all this works without exposing the app to a security vulnerability. Then, you will load this into the Java class and compile and run the class, and hopefully, it works.
```
small tweaks to tighten this up.